lib/cluster/index.ts

```diff
@@ -0,0 +1,646 @@
+        if (err && err.message === 'Failed to refresh slots cache.') {
+          Redis.prototype.silentEmit.call(this, 'error', err)
+          this.connectionPool.reset([])
+        }
+      }.bind(this))
+      this.subscriber.start()
+    })
+  }
+
+  /**
+   * Called when closed to check whether a reconnection should be made
+   *
+   * @private
+   * @memberof Cluster
+   */
+  private handleCloseEvent(): void {
+    let retryDelay
+    if (!this.manuallyClosing && typeof this.options.clusterRetryStrategy === 'function') {
+      retryDelay = this.options.clusterRetryStrategy.call(this, ++this.retryAttempts)
+    }
+    if (typeof retryDelay === 'number') {
+      this.setStatus('reconnecting')
+      this.reconnectTimeout = setTimeout(function () {
+        this.reconnectTimeout = null
+        debug('Cluster is disconnected. Retrying after %dms', retryDelay)
+        this.connect().catch(function (err) {
+          debug('Got error %s when reconnecting. Ignoring...', err)
+        })
+      }.bind(this), retryDelay)
+    } else {
+      this.setStatus('end')
+      this.flushQueue(new Error('None of startup nodes is available'))
+    }
+  }
+
+  /**
+   * Disconnect from every node in the cluster.
+   *
+   * @param {boolean} [reconnect=false]
+   * @memberof Cluster
+   */
+  public disconnect(reconnect: boolean = false) {
+    const status = this.status
+    this.setStatus('disconnecting')
+
+    if (!reconnect) {
+      this.manuallyClosing = true
+    }
+    if (this.reconnectTimeout) {
+      clearTimeout(this.reconnectTimeout)
+      this.reconnectTimeout = null
+      debug('Canceled reconnecting attempts')
+    }
+    if (this.slotsTimer) {
+      clearInterval(this.slotsTimer)
+      this.slotsTimer = null
+    }
+
+    this.subscriber.stop()
+    if (status === 'wait') {
+      this.setStatus('close')
+      this.handleCloseEvent()
+    } else {
+      this.connectionPool.reset([])
+    }
+  }
+
+  /**
+   * Quit the cluster gracefully.
+   *
+   * @param {CallbackFunction<'OK'>} [callback]
+   * @returns {Promise<'OK'>}
+   * @memberof Cluster
+   */
+  public quit(callback?: CallbackFunction<'OK'>): Promise<'OK'> {
+    const status = this.status
+    this.setStatus('disconnecting')
+
+    this.manuallyClosing = true
+
+    if (this.reconnectTimeout) {
+      clearTimeout(this.reconnectTimeout)
+      this.reconnectTimeout = null
+    }
+    if (this.slotsTimer) {
+      clearInterval(this.slotsTimer)
+      this.slotsTimer = null
+    }
+
+    const Promise = PromiseContainer.get()
+    if (status === 'wait') {
+      const ret = asCallback(Promise.resolve('OK'), callback)
+
+      // use setImmediate to make sure "close" event
+      // being emitted after quit() is returned
+      setImmediate(function () {
+        this.setStatus('close')
+        this.handleCloseEvent()
+      }.bind(this))
+
+      return ret
+    }
+    return asCallback(
+      Promise.all(this.nodes().map(function (node) {
+        return node.quit()
```

Comment:
this sometimes would throw if node that we issue quit to is already in the disconnected state (often in the tests where redis instance is down before the quit is processed)

Reply:
Good catch! I'll fix this in a separate commit.